199 spotify pr (#270)
deborahgu/soundcork: docs/spotify.md

> # Spotify on SoundTouch

Reviewer: ```suggestion
# Spotify on SoundTouch

As always, requires a Premium Spotify account to work.
```

## Two Different Spotify Systems

There are two completely separate ways Spotify works on a SoundTouch speaker. This is a common source of confusion.

### 1. Spotify Connect (Always Works)

- The speaker advertises itself as a Spotify Connect device on your local network
- Open the Spotify app on your phone or computer, tap the speaker/device icon, and select your SoundTouch speaker
- Audio streams directly from Spotify's CDN to the speaker
- **No Bose servers involved** — purely between the Spotify app, Spotify's servers, and the speaker

> - **No soundcork involvement** — Spotify Connect operates independently

Reviewer: ```suggestion
- Doesn't use Soundcork in any way.
```

mmm, spicy autocomplete.

### 2. SoundTouch Spotify Integration (Presets)

- This is what the SoundTouch app used for setting Spotify presets (buttons 1-6)
- Originally relied on Bose's OAuth token management via the marge server

> - After the Bose shutdown, this path is broken **unless soundcork handles the token refresh**

Reviewer: ```suggestion

```

> ## Automatic Spotify Support

Reviewer: ```suggestion
## Spotify managed by speakers or Soundtouch software using Soundcork
```

> Soundcork can keep Spotify presets working automatically.

Reviewer: ```suggestion
Soundcork allows users to maintain Spotify presets, and continue to play Spotify streams over any Soundtouch-enabled app.
```

### OAuth Token Intercept (Ongoing Refresh)

> Once the speaker has an active Spotify session (from the ZeroConf primer or a previous Spotify Connect cast), it will periodically refresh its token by calling an OAuth endpoint. Soundcork intercepts these requests and returns a valid token.

Reviewer: ```suggestion
### OAuth Token Intercept (Ongoing Refresh)

The speaker requests a Spotify session from the synthetic Bose APIs (Soundcork), which negotiates for a session token with the Spotify APIs.

Once the speaker has an active Spotify session (from the ZeroConf primer or a previous Spotify Connect cast), it will periodically refresh its token by calling a Bose OAuth endpoint. Soundcork intercepts these requests and returns a valid token.
```

> **Note:** The SoundTouch speakers don't have a separate configuration for the OAuth server. Rather, they take the marge server address and append `oauth` to the end of the first part of the hostname. So for the Bose systems, this this changing `https://streaming.bose.com`  to  `https://streamingoauth.bose.com`. For Soundcork to work with Spotify, it must be available both at a hostname and at an oauth hostname, so `soundcork.local.domain` and `soundcorkoauth.local.domain`.

Reviewer: `this this` should be something?

**How it works:**

1. Speaker sends `POST {oauth_server}/oauth/device/{deviceId}/music/musicprovider/15/token/cs3`
2. Soundcork refreshes the token using the stored Spotify account credentials
3. Returns a fresh access token as JSON
4. The speaker uses this token for continued Spotify playback

> This is transparent — the speaker manages its own refresh cycle, just like it did with `streamingoauth.bose.com`. 

Reviewer: ```suggestion
```

## Setup

### Step 1: Register a Spotify App

1. Go to the [Spotify Developer Dashboard](https://developer.spotify.com/dashboard)
2. Create a new app
   - **Redirect URI**: `{your-soundcork-url}/mgmt/spotify/callback` (e.g., `https://soundcork.local:8000/mgmt/spotify/callback`)
   - **APIs used**: Web API
3. Note the **Client ID** and **Client Secret**

### Step 2: Configure Soundcork

**NOTE: this configuration may change in the future**

> Set the environment variables:

Reviewer: oof.

Using the management API:

```bash
# Start the OAuth flow
curl -u admin:password https://your-soundcork/mgmt/spotify/auth/init

# Open the returned URL in your browser, authorize, then complete:
curl -u admin:password "https://your-soundcork/mgmt/spotify/auth/callback?code=AUTH_CODE"

# Verify the account is linked
curl -u admin:password https://your-soundcork/mgmt/spotify/accounts
```

### Step 4: Add refresh token to source configuration

**TODO**

The refresh token from Spotify must be added to the Source configuration. In the near future a web UI will do this for you. For now, you can edit your Sources.xml file directly.

```
   <source id="34" secret="{your refresh token, should start with AQ}" secretType="token_version_3">
        <sourceKey type="SPOTIFY" account="{your account id, should be a long alphanumeric string}" />
    </source>
```

After the account setup above, both `secret` and `account` information should be available in `{soundcorkdbdir}/spotify/accounts.json`.

### Step 5: Verify

After linking your Spotify account:
- The OAuth intercept works immediately — the speaker will get fresh tokens on its next refresh cycle

> - Press a Spotify preset button on the speaker — it should play

Reviewer: ```suggestion
After linking your Spotify account, press a Spotify preset button on the speaker. If nothing plays, check your server logs.
```

## Technical Details

### Token Lifecycle

> - Spotify access tokens expire after **1 hour** (3600 seconds)

Reviewer: ```suggestion
- Spotify access tokens expire after 1 hour (3600 seconds)
```

- The speaker's firmware requests a new token via the OAuth endpoint before expiry
- Soundcork caches tokens to avoid unnecessary Spotify API calls

### What `cs3` / `token_version_3` Means

> The speaker requests tokens with `tokenType=cs3` in the URL. This corresponds to `token_version_3` in the XML credential format — it's Bose's internal versioning for their OAuth credential schema. The actual value is a standard Spotify access token.

Reviewer: ```suggestion
```
